stream_zip.py

```diff
@@ -612,19 +624,25 @@ def _no_compression_streamed_data(chunks, uncompressed_size, crc_32, maximum_siz
             name_encoded = name.encode('utf-8')
             _raise_if_beyond(len(name_encoded), maximum=0xffff, exception_class=NameLengthOverflowError)
 
+            # Remove time zone information (if any) during clamp
+            mod_datetime_ms_dos = min(max(modified_at.replace(tzinfo=None), _MS_DOS_DATE_BEGIN), _MS_DOS_DATE_END)
             mod_at_ms_dos = modified_at_struct.pack(
-                int(modified_at.second / 2) | \
-                (modified_at.minute << 5) | \
-                (modified_at.hour << 11),
-                modified_at.day | \
-                (modified_at.month << 5) | \
-                (modified_at.year - 1980) << 9,
+                (mod_datetime_ms_dos.second // 2) | \
+                (mod_datetime_ms_dos.minute << 5) | \
+                (mod_datetime_ms_dos.hour << 11),
+                mod_datetime_ms_dos.day | \
+                (mod_datetime_ms_dos.month << 5) | \
+                (mod_datetime_ms_dos.year - 1980) << 9,
             )
             mod_at_unix_extra = mod_at_unix_extra_struct.pack(
                 mod_at_unix_extra_signature,
                 5,        # Size of extra
                 b'\x01',  # Only modification time (as opposed to also other times)
-                int(modified_at.timestamp()),
+                # Clamp timestamp to fit the field size (4-byte signed integer)
+                # In principle, we the lower limit should be `-2**31` but we set it
+                # to zero to avoid issues with common zip utilities like `unzip`.
+                # Said tools do not correctly interpret negative timestamps.
+                max(min(int(modified_at.timestamp()), 2**31 - 1), 0),
```

Comment:
From (admittedly only brief tests) it looks like once times go below 1970 with extended timestamps, then unzip seems to clamp to 1980 (so probably using the MS DOS timestamp?). While odd certainly, I guess I'm leaning to going for the full range in the zip file, because the timestamp before 1970 will be wrong in unzip no matter what, but we can at least make timestamps before 1970 right for other clients.